### ISSUE TITLE
Android in-app purchase crash on some device (Godot 2.1.4)

### ISSUE THREAD

**Issue by megasoft78** [godotengine/godot#17481]
Hi,
I released my little word game on Google Play Store:
[Falling Word](https://play.google.com/store/apps/details?id=it.megasoft78.fallingword)
I also integrated ACRA into the apk to check for crashes.
I'm getting the following stacktrace from ACRA when user try to buy in-app:

```
java.lang.RuntimeException
:Failure delivering result ResultInfo{who=null, request=4097, result=0, data=null} to activity {it.megasoft78.fallingword/org.godotengine.godot.Godot}
:java.lang.NullPointerException
:Attempt to invoke virtual method 'java.lang.String android.content.Intent.getStringExtra(java.lang.String)' on a null object reference
at  android.app.ActivityThread.deliverResults (ActivityThread.java : 4481)
at  android.app.ActivityThread.handleSendResult (ActivityThread.java : 4524)
at  android.app.ActivityThread.-wrap22ActivityThread.java
at  android.app.ActivityThread$H.handleMessage (ActivityThread.java : 1691)
at  android.os.Handler.dispatchMessage (Handler.java : 102)
at  android.os.Looper.loop (Looper.java : 154)
at  android.app.ActivityThread.main (ActivityThread.java : 6692)
at  java.lang.reflect.Method.invokeNative Method
at  com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java : 1468)
at  com.android.internal.os.ZygoteInit.main (ZygoteInit.java : 1358)
Caused by : java.lang.NullPointerException
:Attempt to invoke virtual method 'java.lang.String android.content.Intent.getStringExtra(java.lang.String)' on a null object reference
at  org.godotengine.godot.payments.HandlePurchaseTask.handlePurchaseRequest (HandlePurchaseTask.java : 64)
at  org.godotengine.godot.payments.PaymentsManager.processPurchaseResponse (PaymentsManager.java : 228)
at  org.godotengine.godot.Godot.onActivityResult (Godot.java : 250)
at  android.app.Activity.dispatchActivityResult (Activity.java : 7226)
at  android.app.ActivityThread.deliverResults (ActivityThread.java : 4477))
...  9 more
java.lang.NullPointerException
:Attempt to invoke virtual method 'java.lang.String android.content.Intent.getStringExtra(java.lang.String)' on a null object reference
at  org.godotengine.godot.payments.HandlePurchaseTask.handlePurchaseRequest (HandlePurchaseTask.java : 64)
at  org.godotengine.godot.payments.PaymentsManager.processPurchaseResponse (PaymentsManager.java : 228)
at  org.godotengine.godot.Godot.onActivityResult (Godot.java : 250)
at  android.app.Activity.dispatchActivityResult (Activity.java : 7226)
at  android.app.ActivityThread.deliverResults (ActivityThread.java : 4477)
at  android.app.ActivityThread.handleSendResult (ActivityThread.java : 4524)
at  android.app.ActivityThread.-wrap22ActivityThread.java
at  android.app.ActivityThread$H.handleMessage (ActivityThread.java : 1691)
at  android.os.Handler.dispatchMessage (Handler.java : 102)
at  android.os.Looper.loop (Looper.java : 154)
at  android.app.ActivityThread.main (ActivityThread.java : 6692)
at  java.lang.reflect.Method.invokeNative Method
at  com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java : 1468)
at  com.android.internal.os.ZygoteInit.main (ZygoteInit.java : 1358)
```

This issue doesn't happen all the times and I also tested the in-app purchase on Samsung S5 with a tester account and it's working.
I'm not sure but from what I'm getting, it seems to happen on some Samsung devices.

Could you please help me?

**Comment by akien-mga** [comment]
Please fill in the issue template:

**Godot version:**



**OS/device including version:**

**Comment by megasoft78** [comment]
Hi,
**Godot version**: 2.1.4
**OS/device including version**: 
At the moment I have a crash on the following devices:

- Samsung SM-G930F heroltexx (Android 7.0)

- Samsung SM-T110 goyawifixx (Android 4.2.2)

- Samsung SM-J320F j3xltejt (Android 5.1.1)

**Comment by volzhs** [comment]
looks duplicated #10456
and I found this https://github.com/onepf/OpenIAB/issues/233#issuecomment-45407837

**Comment by volzhs** [comment]
> Is it possible that android:launchMode="singleTask" is the cause of it?

I think so.
it reminds me I had the same issue with another engine a long time ago.

**Comment by megasoft78** [comment]
Do you know why we use that in Godot?

**Comment by volzhs** [comment]
hm. I just searched my old project with another engine and I found I used it also `android:launchMode="singleTask"`
I had a similar issue at that time, but it's been years, can't remember what that was exactly.
according to my old source file, I guess `"singleTask"` is fine.
there should be something another thing.

if i remember correctly, without `"singleTask"`, game can launch in multiple instances.

1. run game - 1 instance
2. got notification of the game while it's running
3. run another game by touching notification - 2 instances

**Comment by megasoft78** [comment]
Oh ok,
It must be something else then. :(

**Comment by volzhs** [comment]
not happen all the time on the same device?

**Comment by megasoft78** [comment]
No,
it's happening random. I list those 3 devices but I can list more if I wait a bit more. :(

**Comment by megasoft78** [comment]
I used libgdx before for other games and I don't see that launchMode in AndroidManifest.xml
It mean is using the default value that is stardard.
How libgdx was able to overcome that issue?

**Comment by volzhs** [comment]
I've never used libgdx, but cocos2d-x.
I had that issue so I had to add it with cocos2d-x.

**Comment by megasoft78** [comment]
Ok,
I'm able to reproduce.
It seems if you click on purchase and then click on home button it will crash.
Now I don't know how to workaround it. Not sure remove singleTask is a good idea :(

**Comment by megasoft78** [comment]
If it's just in this case probably is safe to just put 
> if (data != null) 

at the begin of the method

> public void handlePurchaseRequest(int resultCode, Intent data)

of the class 
> HandlePurchaseTask

What do you think?

**Comment by volzhs** [comment]
@megasoft78 
yeah, data should not be null in any case.
would you test with this code in `platform/android/java/src/org/godotengine/godot/payments/HandlePurchaseTask.java`?

```java
public void handlePurchaseRequest(int resultCode, Intent data) {
    if (data == null) {
        canceled();
        return;
    }
    .....
}
```

**Comment by megasoft78** [comment]
I will but I'm not sure if this is correct.
Is it possible that because of the singleTask and a lack of memory on the device this happen even when the user try to purchase an in-app?
I'm a bit worry that instead of fixing the issue I'm just hiding it. :(

**Comment by volzhs** [comment]
There could be a possibility.
but I guess app would be dead at any moment if app dies because of lack of memory when purchasing.

**Comment by megasoft78** [comment]
Yes, you are right.
I think I'm going to implement that fix. :)

**Comment by volzhs** [comment]
PR is welcome. :smile:

**Comment by megasoft78** [comment]
Yep,
I just need to understand how it works :)

**Comment by megasoft78** [comment]
Hi @volzhs,
I'm not confident with git. :(
The change is very simple:
I just added an if statement in the file HandlePurchaseTask at the method handlePurchaseRequest.
The code looks like this:

> if (data != null) {
>    ...
> }

I hope you can create the PR for me. :)

**Comment by akien-mga** [comment]
Fixed by #17954.
